### PR TITLE
Ships

### LINKED ISSUE #3: Create a way for ships to be defined in raw text
Likely JSON or something similar.

I want a plain text file so that the end users can easily make custom ships or make modifications if they want to experiment.

This will need to have everything represented on the card, AND a way to determine the hull zone geometry.

I don't really care how this gets done, as long as it is readable and easy to implement.  In a prior implementation I used the distance along the side edge for where the lines hit, and a distance where the 2 front arcs met since those were all easy measurements.

The upcoming munificent AND the Onager have arcs that are actually narrower than the front edge of the base, so we will need a different way to indicate that without adding too much overhead.

### PR DESCRIPTION
**Pull request by Sobonis01**

### CODE REVIEW

File: Ships as Raw.txt [hunk 1]
```diff
@@ -0,0 +1,780 @@
+Zoom 93% 
+
```
Comment (danbuis) [inline Ships as Raw.txt]: Not sure what this is, let's remove it.

File: Ships as Raw.txt [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+Points|56
+Hull|5
+Defense token|Br Rd Ev
+Command|2 
+Squad|2 
+Eng|3 
+Nav chart 
+Speed 1|2 
+Speed 2|1 1 
+Speed 3|0 1 1
```
Comment (danbuis) [inline Ships as Raw.txt]: All rows should have key|value. This one doesn't.

We also don't have a Speed 4.  Now you and I know that a gladiator can't get to speed 4.  I'd like to propose 2 solutions here.

1) We can do something like Nav chart:2 11 011 -
It puts all the data in one row, at the expense of more difficulty parsing later.

OR

2) We drop the nav chart line and add in a line for Speed 4 like Speed 4: -

File: Ships as Raw.txt [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+Nav chart 
+Speed 1|1
+Speed 2|1 2
+Front Arc 4 |RRBB
+Side Arc 3 |RRR
+Rear Arc 2 |RR
+Anti Squadron|R
+Upgrade|Title Officers TurboLaser Offensive Support Team Ion Cannon  
+FrontArcOffset| 7.9
+//How far from the front corner does the hull zone line hit the cardboard edge//
```
Comment (danbuis) [inline Ships as Raw.txt]: Muni is weird in that the front arc offset is probably not how far from the front corner since it actually hits along the front edge instead of the side.

Maybe a negative value and when we parse we can act accordingly?

File: Ships as Raw.txt [hunk 4]
```diff
@@ -0,0 +1,780 @@
...
+//How far from the rear corner does the hull zone line hit the cardboard edge//
+FrontArcConjunction|53
+//How far from the front of the ship do the front hull zone lines meet//
+RearArcConjunction|49
+//How far from the rear of the ship do the rear hull zone lines meet//
+
+
+Name|Munificent-class Star Frigate
+Keyword| Droid Bombard Transport
+Faction|SA
```
Comment (danbuis) [inline Ships as Raw.txt]: I added some factions to the squadron parses PR, if we can match those that would be awesome.

I can absolutely be talked into kicking this one down the road for now though.

File: Ships as Raw.txt [hunk 5]
```diff
@@ -0,0 +1,780 @@
...
+FrontArcOffset|15.1
+//How far from the front corner does the hull zone line hit the cardboard edge//
+RearArcOffset|5.2
+//How far from the rear corner does the hull zone line hit the cardboard edge//
+FrontArcConjunction|41
+//How far from the front of the ship do the front hull zone lines meet//
+RearArcConjunction|29
+//How far from the rear of the ship do the rear hull zone lines meet//
+
+Name|Acclamator II
```
Comment (danbuis) [inline Ships as Raw.txt]: I'm fairly sure the card says something like assault cruiser in addition to this.

File: Ships as Raw.txt [hunk 6]
```diff
@@ -0,0 +1,780 @@
...
+Eng|2
+Nav chart 
+Speed 1|2
+Speed 2|2 1 
+Speed 3|1 1 1
+Speed 4|1 0 1 1 
+Front Arc 2 |BBK
+Side Arc 2 |BK
+Rear Arc 1 |B
+Anti Squadron|bb
```
Comment (danbuis) [inline Ships as Raw.txt]: These should probably be upper case

**Review by danbuis — CHANGES_REQUESTED** [review]
A few change requests.  I didn't do a super thorough look yet, but this is enough to get started.

Some proper parsing validation should weed out everything else that isn't just bad copying from the cards.

File: Ships as Raw.txt (hunk 3 above)
Comment (danbuis) [reply Ships as Raw.txt]: Marking this as resolved to be handled in a specifically targetted issue.

File: Ships as Raw.txt (hunk 2 above)
Comment (Sobonis01) [reply Ships as Raw.txt]: Lets do number 2. I would say lets try 
Speed 1|2 
Speed 2|1 1 
Speed 3|0 1 1
Speed 4|-

**Comment by Sobonis01** [comment]
[Ships as Raw.txt](https://github.com/danbuis/ArmadaAIEngineNew/files/5817590/Ships.as.Raw.txt)

**Comment by Sobonis01** [comment]
[Ships as Raw.txt](https://github.com/danbuis/ArmadaAIEngineNew/files/5817661/Ships.as.Raw.txt)

**Comment by danbuis** [comment]
2 real quick github things, unrelated to your edits.

1) Its customary to let the reviewer mark a conversation as resolved.
2) You don't need to attach the file in a comment here, as once you make a commit locally and then push it the new file is available for the rest of us.

File: Ships as Raw.txt [hunk 7]
```diff
@@ -70,10 +70,10 @@ Speed 1|2
...
 Speed 3|1 1 1
 Speed 4|1 0 1 1 
-Front Arc 2 |RRB
-Side Arc 2 |RR
-Rear Arc 1 |RB
-Anti Squadron|K
+FrontArc 2 |RRB
+SideArc 2 |RR
+RearArc 1 |RB
+AntiSquadron|K
```
Comment (danbuis) [inline Ships as Raw.txt]: I like the CamelCaseing here instead of 2 words.  Let's carry that elsewhere so that all the keys are 1 word, ie like lines 43-45.

File: Ships as Raw.txt [hunk 8]
```diff
@@ -159,10 +160,11 @@ Nav chart
...
 Speed 2|1 0 
 Speed 3|0 1 0
-Front Arc 4 |RRRKKK
-Side Arc 2 |RRK
-Rear Arc 2 |RBK
-Anti Squadron|KK
+Speed 4|-
+FrontArc 4 |RRRKKK
+SideArc 2 |RRK
+RearArc 2 |RBK
```
Comment (danbuis) [inline Ships as Raw.txt]: You mentioned this in another issue, let's get these all changed to whatever we decide in that conversation thread.

File: Ships as Raw.txt [hunk 9]
```diff
@@ -637,7 +656,7 @@ RearArcConjunction|62
 
 Name|Imperial Star Destroyer Kuat Refit
```
Comment (danbuis) [inline Ships as Raw.txt]: This plus line 686, looks suspiciously like a repeat.  Am I correct, or did we just miss renaming this entry?

File: Ships as Raw.txt [hunk 10]
```diff
@@ -177,7 +179,7 @@ RearArcConjunction|41
 51++
 Name|Victory I-class Star Destroyer
```
Comment (danbuis) [inline Ships as Raw.txt]: It would ne *nice* if the ship entries were grouped by faction.  We start with Imperial, float through the prequal factions, and then came back to Imperial.

**Review by danbuis — DISMISSED** [review]
Some more minor change requests.

Let's also make sure that this file is going to the same directory as the squadrons text file.  ie something like assets/data/thisFile.txt

File: The Shipyard.txt [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+
+package gameComponents;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+
+public class Ships {
+public class Faction {
+
```
Comment (danbuis) [inline The Shipyard.txt]: A few things happening here.

1) Did you intend to include this new file here in the same PR as the ship text?  I had expected they would be separate branches and separate pull requests so that when the text file was done it could be merged.  If these 2 stay in the same pull request/branch, *which is fine* then they won't be merged for a while.

2) I'm guessing you didn't make this file in your IDE, its missing a few closing braces, and nesting classes like this is weird.  Was that intentional?  If so, what was the reasoning?

3) Wrong file name/extension.  I'm about 95% sure that the class and file name should be the same, ie if this file defines the Ships class, the file should be Ships.java, or if it is the ShipFactory class it would be ShipFactory.java.  IntelliJ should do that automatically for you.   Just right click on the folder you want your class -> New -> Java Class, and type in your name.

File: The Shipyard.txt [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    private final String size;
+    private final String name;
+    private final String faction;
+    private final int Hull;
```
Comment (danbuis) [inline The Shipyard.txt]: Don't capitalize variables.  Its not a requiremnt of the language, but it is convention because it makes it a lot easier to read code later on.  Objects are given capital names, ie Ship, HullZone, Squadron, String.  If you have a capitalized variable, it looks like an object type.  If I violated that guideline in the corresponding SquadronFactory class, please feel free to call it out.

Comment (danbuis) [reply The Shipyard.txt]: This is particularly important on your `Upgrade` fields here.  We are likely to have an `Upgrade` object at some point that is interacting with the `Ship` objects.

It is common convention to include the type of the variable or to err on the side of overly descriptive.  For instance squadValue, so that we don't get confused with the Squadron object, or upgradeList so that we know what kind of object it is.

File: Ships as Raw.txt [hunk 13]
```diff
@@ -654,151 +530,247 @@ FrontArcConjunction|65
...
+Squad|2
+Eng|3
+Nav chart
+Speed 1|2
+Speed 2|1 1
+Speed 3|1 1 0
+Speed 4|-
+FrontShields|3 FrontArmament|RRR
+SideShields|1 SideArmament|R
+RearShields|2 RearArmament|-
```
Comment (danbuis) [inline Ships as Raw.txt]: These 3 lines should be 6.

ie 
FrontShields|3
FrontArmament|RRR

Since we can't have more than 1 "pipe" (`|`) per line

Pretty sure I saw this in a few other spots

File: The Shipyard.txt (hunk 12 above)
Comment (Sobonis01) [reply The Shipyard.txt]: Also I am going to use the squadron builder version of defenseTokens, so it is the same for both.

**Review by Sobonis01 — COMMENTED** [review]
will do.

File: Ships as Raw.java.txt [hunk 14]
```diff
@@ -0,0 +1,881 @@
+
+Name|Gladiator I-class Star Destroyer
+Type|Gladiator
```
Comment (danbuis) [inline Ships as Raw.java.txt]: Might want to check the file name on this guy.

This looks like a duplicate file, and has a .java. tossed in the middle

**Review by Sobonis01 — COMMENTED** [review]
updated as raw

**Review by Sobonis01 — COMMENTED** [review]
Made changes to ship file.

File: The Shipyard.txt (hunk 11 above)
Comment (Sobonis01) [reply The Shipyard.txt]: Let me make this a separate thing

**Comment by danbuis** [comment]
Is this one ready for review at this point?

**Comment by Sobonis01** [comment]
Yes it is the one for review.

**Comment by danbuis** [comment]
Cool, let's toss a "ready for review" label on it.  You can find it under the labels dropdown on the left side of the screen.

I'll toss it on here and take a look here in a bit.

File: Ships as Raw [hunk 15]
```diff
@@ -0,0 +1,860 @@
+
```
Comment (danbuis) [inline Ships as Raw]: I'd like a better name on this file, preferably matching the squadron one, so something like ships.txt.

Let's make sure that its also in the assets folder so that all our data files live in the same place.

File: Ships.java [hunk 16]
```diff
@@ -0,0 +1,862 @@
+
```
Comment (danbuis) [inline Ships.java]: What is the intended use for this file?  It isn't formatted like a java file and looks identical to the text file

File: The Shipyard.txt [hunk 17]
```diff
@@ -0,0 +1,58 @@
+
```
Comment (danbuis) [inline The Shipyard.txt]: Since this funcitonality is being done in a different PR, we can probably remove this file from here.  It clutters things up and contributes little to the intended goal of this PR.

Plus its a text file that looks like it is trying to be a poorly formatted java file.

File: Ships as Raw [hunk 18]
```diff
@@ -0,0 +1,860 @@
...
+AntiSquadron|RK
+Upgrade|Title Officers TurboLaser Defensive Suport Team Ion Cannon
+FrontArcOffset| 7.9
+//How far from the front corner does the hull zone line hit the cardboard edge//
+RearArcOffset|3.2
+//How far from the rear corner does the hull zone line hit the cardboard edge//
+FrontArcConjunction|53
+//How far from the front of the ship do the front hull zone lines meet//
+RearArcConjunction|49
+//How far from the rear of the ship do the rear hull zone lines meet//
```
Comment (danbuis) [inline Ships as Raw]: I'm not wild about the comments on the geometry lines being on each instance of the ship.

Definitely have them on the first one or 2.  

Not concerned enough to ask you to remove them though if you don't want to.  Just registering that I'm not to keen on them.

File: Ships as Raw [hunk 19]
```diff
@@ -0,0 +1,860 @@
...
+//How far from the rear corner does the hull zone line hit the cardboard edge//
+FrontArcConjunction|65
+//How far from the front of the ship do the front hull zone lines meet//
+RearArcConjunction|62
+//How far from the rear of the ship do the rear hull zone lines meet//
+
+
+Name|MC75 Ordnance Cruiser
+Type|MC75 Ordnance Cruiser
+Keywords|0
```
Comment (danbuis) [inline Ships as Raw]: What was the rationale on the 0 vs None again?  I def did None on the squadrons, which isn't to say that it is better than 0.

Mostly a fan of consistancy.

File: Ships as Raw [hunk 20]
```diff
@@ -0,0 +1,860 @@
...
+//This ship has a 2 section between the rear and front conjuctions.//
+
+Name|Assault Frigate Mark II B
+Type|Assault Frigate Mark II
+Keywords|0
+Faction|Rebel
+Size|Medium
+Points|72
+Hull|6
+defenseTokens|Rd Ev Br
```
Comment (danbuis) [inline Ships as Raw]: I'd like defenseTokens to be DefenseTokens, with the capital at the front

File: Ships as Raw [hunk 21]
```diff
@@ -0,0 +1,860 @@
...
+Command|3
+Squad|2
+Eng|4
+Speed1|1
+Speed2|11
+Speed3|011
+Speed4|0
+Anti Squadron|B
```
Comment (danbuis) [inline Ships as Raw]: Lets remove the space to maintain CamelCasing, like FrontShield and such

File: Ships as Raw (hunk 20 above)
Comment (danbuis) [reply Ships as Raw]: Looks like this applies to most entries

File: Ships as Raw [hunk 22]
```diff
@@ -0,0 +1,860 @@
...
+Speed3|011
+Speed4|0
+FrontShields|3
+SideShields|2
+RearShields|1
+FrontArmament|RRKK
+SideArmament|KKKK
+RearArmament|RK
+AntiSquadron|B
+Upgrade|Title Officers Weapon Team Support Ordance
```
Comment (danbuis) [inline Ships as Raw]: Remove the space in WeaponTeam, or remove Team.  Basically lets have as many words as upgrades.  I think you have an extra word here.

File: Ships as Raw [hunk 23]
```diff
@@ -0,0 +1,860 @@
...
+Speed3|011
+Speed4|0
+FrontShields|3
+SideShields|2
+RearShields|1
+FrontArmament|RRKK
+SideArmament|KKKR
+RearArmament|RK
+AntiSquadron|BB
+Upgrade|Title Officers Weapon Team Support Ordance
```
Comment (danbuis) [inline Ships as Raw]: I'd prefer Officer over Officers, since its one officer for the slot, and that's the name given in the wiki, and maybe even the rulebook.  but I'm willing to not push to hard on that.

File: Ships as Raw [hunk 24]
```diff
@@ -0,0 +1,860 @@
...
+Speed3|0
+Speed4|0
+FrontShields|3
+SideShields|3
+RearShields|1
+FrontArmament|RRRKKK
+SideArmament|KRRR
+RearArmament|RR
+AntiSquadron|B
+Upgrade|Title Officers TurboLaser Offensive Weapon Team Ordance
```
Comment (danbuis) [inline Ships as Raw]: Ordnance, not Ordance

File: Ships as Raw [hunk 25]
```diff
@@ -0,0 +1,860 @@
...
+Speed3|0
+Speed4|0
+FrontShields|3
+SideShields|3
+RearShields|1
+FrontArmament|RRRBBB
+SideArmament|BRR
+RearArmament|RR
+AntiSquadron|B
+Upgrade|Title Officers TurboLaser Offensive Weapon Team Ion cannon
```
Comment (danbuis) [inline Ships as Raw]: Turbolaser, not TurboLaser.  I *think* its one word, but my lack of SW knowledge might be failing me

File: Ships as Raw [hunk 26]
```diff
@@ -0,0 +1,860 @@
...
+Speed3|011
+Speed4|0
+FrontShields|4
+SideShields|3
+RearShields|2
+FrontArmament|RRRRRBB
+SideArmament|RBBB
+RearArmament|RBB
+AntiSquadron|KK
+Upgrade|Title Officers Offensive Weapons Team Turbo Laser Turbo Laser Fleet Command
```
Comment (danbuis) [inline Ships as Raw]: Turbo Laser should definitely be one thing, not 2.  Ditto for FleetCommand

File: Ships as Raw [hunk 27]
```diff
@@ -0,0 +1,860 @@
...
+Speed3|011
+Speed4|0
+FrontShields|4
+SideShields|3
+RearShields|2
+FrontArmament|RRRBBKKK
+SideArmament|RRKK
+RearArmament|RBB
+AntiSquadron|BK
+Upgrade|Title Officers Defensive Offensive Offensive Weapons Team Turbo Laser Defensive Ion Cannon
```
Comment (danbuis) [inline Ships as Raw]: IonCannon, WeaponsTeam, Turbolaser, etc.  change to be consistant with whatever you choose.  At the end of the day each card should get one token, or string of chars between spaces.

**Review by danbuis — CHANGES_REQUESTED** [review]
Comments to improve consistancy and to be compatible with the existing parsing utils that you will be soon using.

Maybe some deadweight in unneeded files.

File: Ships as Raw [hunk 28]
```diff
@@ -0,0 +1,860 @@
...
+Points|56
+Hull|5
+defenseTokens|Br Rd Ev
+Command|2 
+Squad|2 
+Eng|3 
+Speed1|2
+Speed2|11
+Speed3|011
+Speed4|0
```
Comment (danbuis) [inline Ships as Raw]: Probs want a different way to note an invalid speed, since we have 0 as 0-clicks elsewhere

Comment (Sobonis01) [reply Ships as Raw]: Should I put a -  sign?

File: Ships as Raw (hunk 15 above)
Comment (Sobonis01) [reply Ships as Raw]: I can change that in the morning

File: Ships.java (hunk 16 above)
Comment (Sobonis01) [reply Ships.java]: Will delete

File: The Shipyard.txt (hunk 17 above)
Comment (Sobonis01) [reply The Shipyard.txt]: I agree let's remove this file

File: Ships as Raw (hunk 18 above)
Comment (Sobonis01) [reply Ships as Raw]: This was in there to act as a guide if someone wanted to make there own ships.  So thought I would describe what they meant.

File: Ships as Raw (hunk 19 above)
Comment (Sobonis01) [reply Ships as Raw]: No problem I didn't think of making them the same I get it

File: Ships as Raw (hunk 28 above)
Comment (danbuis) [reply Ships as Raw]: That sounds good to me.

**Comment by Sobonis01** [comment]
[Ships.java.txt](https://github.com/danbuis/ArmadaAIEngineNew/files/6329201/Ships.java.txt)

File: Ships.java (hunk 16 above)
Comment (danbuis) [reply Ships.java]: You deleted the wrong file.  You can probably fix it by renaming the extension on this one to .txt instead of java.

**Comment by Sobonis01** [comment]
[Ships.java.txt](https://github.com/danbuis/ArmadaAIEngineNew/files/6369669/Ships.java.txt)

**Comment by Sobonis01** [comment]
[Ships.txt](https://github.com/danbuis/ArmadaAIEngineNew/files/6369688/Ships.txt)
